src/jvm/main/org/jetbrains/kotlinx/lincheck/verifier/LTS.kt

```diff
@@ -121,19 +121,19 @@ class LTS(private val sequentialSpecification: Class<*>) {
             }
         }
 
-        private fun Result.isLegalByRequest(transitionInfo: TransitionInfo, allowExtraSuspension: Boolean) =
-            isLegalByFollowUp(transitionInfo, allowExtraSuspension) ||
-            this.wasSuspended && (transitionInfo.result == Suspended || allowExtraSuspension) ||
+        private fun Result.isLegalByRequest(transitionInfo: TransitionInfo) =
+            isLegalByFollowUp(transitionInfo) ||
+            this.wasSuspended ||
             !this.wasSuspended && transitionInfo.result == Suspended
 
-        private fun Result.isLegalByFollowUp(transitionInfo: TransitionInfo, allowExtraSuspension: Boolean) =
+        private fun Result.isLegalByFollowUp(transitionInfo: TransitionInfo) =
             this == transitionInfo.result ||
             this is ValueResult && transitionInfo.result is ValueResult && this.value == transitionInfo.result.value &&
-                (!wasSuspended && transitionInfo.result.wasSuspended || wasSuspended && allowExtraSuspension) ||
+                (!wasSuspended && transitionInfo.result.wasSuspended || wasSuspended) ||
             this is ExceptionResult && transitionInfo.result is ExceptionResult && this.tClassCanonicalName == transitionInfo.result.tClassCanonicalName &&
-                (!wasSuspended && transitionInfo.result.wasSuspended || wasSuspended && allowExtraSuspension) ||
+                (!wasSuspended && transitionInfo.result.wasSuspended || wasSuspended) ||
```

Comment:
I don't think you need the `wasSuspended` marker anymore {---} operations either complete with a result, get canceled, or get suspended and never finish.